tests/integration/relations/pgbouncer_provider/test_pgbouncer_provider.py

```diff
@@ -0,0 +1,339 @@
+                        "upstream-source"
+                    ]
+                },
+                application_name=PGB,
+                num_units=2,
+            ),
+            ops_test.model.deploy(
+                PG,
+                application_name=PG,
+                num_units=2,
+                channel="edge",
+                trust=True,
+            ),
+        )
+        await ops_test.model.add_relation(f"{PGB}:{BACKEND_RELATION_NAME}", f"{PG}:database")
+        await ops_test.model.wait_for_idle(raise_on_blocked=True)
+        # Relate the charms and wait for them exchanging some connection data.
+        relation = await ops_test.model.add_relation(
+            f"{CLIENT_APP_NAME}:{FIRST_DATABASE_RELATION_NAME}", PGB
+        )
+
+    await ops_test.model.wait_for_idle(apps=APP_NAMES, status="active", raise_on_blocked=True)
+
+    client_unit_name = f"{CLIENT_APP_NAME}/0"
+    dbname = "application_first_database"
+
+    # Check we can update and delete things
+    update_query = (
+        "DROP TABLE IF EXISTS test;"
+        "CREATE TABLE test(data TEXT);"
+        "INSERT INTO test(data) VALUES('some data');"
+        "SELECT data FROM test;"
+    )
+    run_update_query = await run_sql_on_application_charm(
+        ops_test,
+        unit_name=client_unit_name,
+        query=update_query,
+        dbname=dbname,
+        relation_id=relation.id,
+    )
+    assert "some data" in json.loads(run_update_query["results"])[0]
+
+    # Check version is accurate
+    version_query = "SELECT version();"
+    run_version_query = await run_sql_on_application_charm(
+        ops_test,
+        unit_name=client_unit_name,
+        query=version_query,
+        dbname=dbname,
+        relation_id=relation.id,
+    )
+    # Get the version of the database and compare with the information that
+    # was retrieved directly from the database.
+    version = await get_application_relation_data(
+        ops_test, CLIENT_APP_NAME, FIRST_DATABASE_RELATION_NAME, "version"
+    )
+    logging.info(run_version_query)
+    logging.info(version)
+    assert version in json.loads(run_version_query["results"])[0][0]
+
+    # Check we can read things in readonly
+    select_query = "SELECT data FROM test;"
+    run_select_query_readonly = await run_sql_on_application_charm(
+        ops_test,
+        unit_name=client_unit_name,
+        query=select_query,
+        dbname=dbname,
+        relation_id=relation.id,
+        readonly=True,
+    )
+    assert "some data" in json.loads(run_select_query_readonly["results"])[0]
+
+    # check we can't write in readonly
+    drop_query = "DROP TABLE test;"
+    run_drop_query_readonly = await run_sql_on_application_charm(
+        ops_test,
+        unit_name=client_unit_name,
+        query=drop_query,
+        dbname=dbname,
+        relation_id=relation.id,
+        readonly=True,
+    )
+    assert "no results to fetch" in json.loads(run_drop_query_readonly["results"])
```

Comment:
Wouldn't be better to have an error being caught by, for example, `pytest.raises`, as the output `no results to fetch` is already returned when some queries succeed (like the ones below about creating database and user)?

Reply:
Because this is running an action, it's actually a bit tricky to expect specific errors with `pytest.raises`, since there's no error from the test's perspective. To get around this, I've rewritten the action to return the error in its `results` field, and it's now checking that we're throwing the correct errors for the correct reason.